src/ray/rpc/worker/core_worker_client.h

```diff
@@ -25,52 +25,97 @@ const int64_t kMaxBytesInFlight = 16 * 1024 * 1024;
 const int64_t kBaseRequestSize = 1024;
 
 /// Get the estimated size in bytes of the given task.
-const static int64_t RequestSizeInBytes(const DirectActorAssignTaskRequest &request) {
+const static int64_t RequestSizeInBytes(const PushTaskRequest &request) {
   int64_t size = kBaseRequestSize;
   for (auto &arg : request.task_spec().args()) {
     size += arg.data().size();
   }
   return size;
 }
 
+/// Abstract client interface for testing.
+class CoreWorkerClientInterface {
```

Comment:
it's probably better to have this class as a virtual base class, aka having all virtual functions = 0 instead of providing a default dummy implementation?

Reply:
This is more convenient for testing.

Reply:
This is not a big deal and I feel it acceptable either way, but usually for testing it's ok to implement some dummy functions, while the interface design is more important.

Reply:
Yeah, strictly speaking the interface should have all abstract methods, then we can have a skeleton with defaults, and tests can inherit from that.

Reply:
Can you instead have all virtual functions = 0, then use gmock to mock the not implemented return values? I think this is standard.

Reply:
Also, can you make the virtual interfaces as minimal as possible? This way it is clear what the dependency injection is, e.g., CoreWorkerDirectTaskSubmitter shouldn't be able to call PushActorTask.

Reply:
The reason I chose to declare all the functions is to make future unit tests as easy as possible, since we should have more of them. I don't think there is any downside of declaring the functions to raise NotImplemented.

Reply:
I think there is a downside, that you can't catch notimplemented errors until runtime, whereas the compiler would normally do it for you.

I don't think it's a big deal to have to declare more interfaces for future unit tests. Actually, I think it's a good thing because then it'll be easier to understand what the dependencies between classes are.